### ISSUE TITLE
Allow users to swap mark positions just as with default motions

### ISSUE THREAD

**Issue by asmodeus812** [vim/vim#15316]
### Steps to reproduce

```vim
set nocompatible

" onoremap i" :<c-u>call <SID>TestInsideObject()<CR>
nnoremap gs :<C-U>set opfunc=<SID>TestInsideQuotes<CR>g@

function! s:TestInsideObject()
    let m1 = getpos(".")
    let m2 = getpos(".")

     let m1[1] = 1
     let m2[1] = 1

     let m1[2] = 2
     let m2[2] = 1

     call setpos("'[", m1)
     call setpos("']", m2)
endfunction

function! s:TestInsideQuotes(motion)
     let m1 = getpos("'[")
     let m2 = getpos("']")

     echom m1
     echom m2
endfunction
```

Have the following file

```txt
line with some text and empty quotes ""
```

1. vim -u config.vim test.txt
2. open the file on the first line
3. Hit `gsi"`

So in this example we have two cases, let us run the repro first without the custom `onoremap` the echo calls would echo out the change marks, where the change marks would actually be reversed, the column more specifically for the start mark `[` will be bigger than the column value for the `]` mark. 

**Which some plugins actually use to identify edge case motions such as this one - where the `inner` part of the text object is emtpy, however you might want to put some text inbetween the quotes, however if the marks' columns had not been reversed you would have no way of knowing you are in the edge case. This is super useful**, however if we add the custom `onoremap`, where we manually set the marks now, we can not force set these marks, after the calls to setpos, vim would then override them, and after we execute `gsi"` and hit `:marks` we see that the values are not the ones we set. 

```txt
" Without onoremap, the values of the marks [ and ], is such that x > y if we do it inside an empty pair of quotes, if the quotes have text in between them the column values are sch that x < y.
[0, 1, x, 0]
[0, 1, y, 0]

" With the onoremap, where we are settings the same marks, these are the values that get printed for me, which do not even correspond to what we have set
[0, 1, 1, 0]
[0, 1, 1, 0]
```

### Expected behaviour

The user should be allowed to replicate this inversed mark behaviour, for his own custom text object, which the default built in motions provide for empty object such as for example quote pairs. Otherwise there is no way to detect the special case where the text object has an `inside` motion that has to capture the content between `opening` and `closing` tags (such as quotes, backticks, whatever) and the content between the tags is empty meaning the text looks like this `{tag}{tag}` and not like this `{tag}{text}{tag}`.

### Version of Vim

9.1

### Environment

Ubuntu22
Alacritty
zsh

### Logs and stack traces

```shell
None
```

**Comment by benknoble** [comment]
[This has been discussed on StackExchange.](https://vi.stackexchange.com/q/45244/10604)

**Comment by asmodeus812** [comment]
Yeah, i am tyring to get more information from knowleadgable people, if it turns out it is not a bug and achieveable without changes to vim itself.

**Comment by benknoble** [comment]
Understood! Having the cross-links helps future readers get the information they're after :)

**Comment by romainl** [comment]
Even better: keep the whole discussion in one place.

**Comment by benknoble** [comment]
> Even better: keep the whole discussion in one place.

I agree, though sometimes it's not feasible (limits of one forum vs another, etc.).

**Comment by asmodeus812** [comment]
So i believe this to be a more convoluted problem. What i am trying to achieve is have onoremap that has no visual select, rather somehow provide the range with other means, or if it does work with visual select it does not register the `mode` in the operator func as `visual` or getmode does not return `visual` since it messes up with the way my operator function works. I am trying to have the onoremap work purely in normal mode, to set the `range` and not trigger any visual selection, not sure how that can be done ?

**Comment by romainl** [comment]
The core issue with your idea is that both text objects and custom operator-pending mode mappings are currently implemented via visual mode. There is no built-in way around that.

The only options available to you seem to be:
- build a whole alternative system that doesn't use visual mode and publish it as a plugin,
- write the necessary code and make a PR.

**Comment by asmodeus812** [comment]
@romainl not quite, the core problem is that i am not sure how to create an operator pending mapping that does not use visual selection to tell vim which range to operate on. As of now if you do not do a visual selection in operator pending mode vim takes the cursor position before the operation starts, and in your operation pending mapping you set the cursor to another position, vim would then operate on these two pairs. If you do a visual select, vim would instead operate on the visual selected range, ignoring the starting cursor position. That is why you can actually create a no-op operator pending mapping by never moving the cursor or doing no visual selection, however this is not helpful to me.

The default builtins seem to be working without doing a visual selection, but a user can not reproduce this. @romainl or can we ?
